### PR TITLE
Allow specification of network announce protocols in Publisher

### PR DESCRIPTION
**Pull request by eile**

### CODE REVIEW

File: zeq/publisher.h [hunk 1]
```diff
@@ -28,8 +27,10 @@ class Publisher : public boost::noncopyable
      * Create a publisher on the given URI.
      *
      * @param uri publishing URI in the format scheme://[*|host|IP|IF][:port]
+     * @param announceMode bitwise combination of AnnounceMode network protocols
      */
-    ZEQ_API explicit Publisher( const lunchbox::URI& uri );
+    ZEQ_API explicit Publisher( const lunchbox::URI& uri,
```
Comment (tribal-tec) [inline zeq/publisher.h]: no more explicit

**Comment by tribal-tec** [comment]
+1

File: zeq/publisher.h (hunk 1 above)
Comment (eile) [inline zeq/publisher.h]: done

**Comment by eile** [comment]
Retest this please

**Comment by eile** [comment]
retest this please

**Comment by tribal-tec** [comment]
Missing rebase onto master

**Comment by tribal-tec** [comment]
retest this please

File: CMakeLists.txt [hunk 2]
```diff
@@ -13,6 +13,8 @@ set(VERSION_PATCH "0")
 set(VERSION_ABI 2)
 
 include(Common)
+include(CTest)
```
Comment (tribal-tec) [inline CMakeLists.txt]: why now? was always part of Common imo.

File: zeq/publisher.h [hunk 3]
```diff
@@ -44,6 +45,9 @@ class Publisher
 
     const std::string& getAddress() const; //!< @internal
 
+    /** @return the publisher URI given in the constructor. */
+    ZEQ_API const lunchbox::URI& getURI() const;
```
Comment (tribal-tec) [inline zeq/publisher.h]: where needed? If really needed, add unit test to keep code coverage high.

File: zeq/publisher.cpp [hunk 4]
```diff
@@ -94,13 +95,32 @@ class Publisher
...
 
 private:
-    void _initService( std::string host, uint16_t port )
+    void _initService( const uint32_t announceMode, std::string host,
+                       uint16_t port )
     {
         _initAddress( host, port );
-        if( lunchbox::Servus::isAvailable( ))
-            _service.announce( port, _address );
+        if( announceMode == ANNOUNCE_NONE || !(announceMode&ANNOUNCE_ZEROCONF) )
```
Comment (tribal-tec) [inline zeq/publisher.cpp]: check for NONE bit? If mode bits are set to ZEROCONF|NONE (as a result of automated code), what should happen?

File: zeq/publisher.h (hunk 3 above)
Comment (hernando) [inline zeq/publisher.h]: I made the same comment yesterday but I don't find it anywhere. Where did it go...?

File: CMakeLists.txt (hunk 2 above)
Comment (eile) [inline CMakeLists.txt]: Test artefact related to Eyescale/CMake#309. Will remove.

File: zeq/publisher.cpp (hunk 4 above)
Comment (eile) [inline zeq/publisher.cpp]: ZEROCONF|NONE == ZEROCONF

Same as 5|0 == 5.

The first check is actually redundant here. Will remove.

File: zeq/publisher.h (hunk 3 above)
Comment (eile) [inline zeq/publisher.h]: Needed in fivox::Compositor. Will add unit test.

**Comment by eile** [comment]
All done

**Comment by eile** [comment]
retest this please

**Comment by eile** [comment]
Retest this please.

File: zeq/publisher.cpp [hunk 5]
```diff
@@ -93,21 +94,46 @@ class Publisher
...
+            uri.getPort(), getAddress( ));
+
+        if( required && !result )
+        {
+            LBTHROW( std::runtime_error( "Zeroconf announce failed: " +
+                                         result.getString( )));
+        }
     }
 
-    void _initAddress( std::string& host, uint16_t& port )
```
Comment (hernando) [inline zeq/publisher.cpp]: The changes you've made in this file, and particularly this function, are going to be very conflicting with hernando:optional_lunchbox. I'd be easier if Publisher::getURI returned a std::string.

**Comment by tribal-tec** [comment]
retest this please

**Comment by tribal-tec** [comment]
retest this please

**Comment by tribal-tec** [comment]
retest this please

**Pull request merged**